torch/csrc/distributed/c10d/reducer.cpp

```diff
@@ -128,36 +130,46 @@ Reducer::Reducer(
         [=](std::vector<int64_t>& v) { v.resize(variable_count); });
   }
 
-  // Initialize locally used parameter maps
-  {
-    const auto replica_count = replicas_.size();
-    const auto variable_count = replicas_[0].size();
-    local_used_maps_.resize(replica_count);
-    local_used_maps_dev_.resize(replica_count);
+  // See Note [Skip allreducing local_used_maps_dev]
+  if (find_unused_parameters_) {
+    // Initialize locally used parameter maps
+    {
+      const auto replica_count = replicas_.size();
+      const auto variable_count = replicas_[0].size();
+      local_used_maps_.resize(replica_count);
+      local_used_maps_dev_.resize(replica_count);
 
-    for (size_t i = 0; i < replica_count; i++) {
-      at::TensorOptions options;
-      options = options.dtype(at::kInt);
+      for (size_t i = 0; i < replica_count; i++) {
+        at::TensorOptions options;
+        options = options.dtype(at::kInt);
+
+        if (replicas_[i][0].is_cuda()) {
+          at::DeviceGuard g(replicas_[i][0].device());
+          local_used_maps_[i] = at::zeros(
+              {static_cast<long>(variable_count)}, options.pinned_memory(true));
+        } else {
+          local_used_maps_[i] =
+              at::zeros({static_cast<long>(variable_count)}, options);
+        }
 
-      if (replicas_[i][0].is_cuda()) {
-        at::DeviceGuard g(replicas_[i][0].device());
-        local_used_maps_[i] = at::zeros(
-            {static_cast<long>(variable_count)}, options.pinned_memory(true));
-      } else {
-        local_used_maps_[i] =
-            at::zeros({static_cast<long>(variable_count)}, options);
+        // This tensor needs to be on the same device as replica because backend
+        // such as NCCL may not support CPU tensors, and hence it might not work
+        // if we always put it on CPU.
+        options = options.device(replicas_[i][0].device());
+        local_used_maps_dev_[i] =
+            at::empty({static_cast<long>(variable_count)}, options);
       }
-
-      // This tensor needs to be on the same device as replica because backend
-      // such as NCCL may not support CPU tensors, and hence it might not work
-      // if we always put it on CPU.
-      options = options.device(replicas_[i][0].device());
-      local_used_maps_dev_[i] =
-          at::empty({static_cast<long>(variable_count)}, options);
     }
   }
 }
 
+// Note [Skip allreducing local_used_maps_dev]
```

Comment:
Just curious, any reason for adding this note to the dtor?

Reply:
In Pritam's example, the notes were placed after the constructor. Thats why I put it there. 
https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/engine.cpp#L100. Example of referring to a note: https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/engine.cpp#L64.